tests/Unit/AutoRouteManagerTest.php

```diff
@@ -11,154 +11,360 @@
-            $this->adapter->generateAutoRouteTag($this->{$contextVar}->reveal())->willReturn('fr');
-            $this->adapter->createAutoRoute($this->{$contextVar}, $this->subject, 'fr')->willReturn($this->{$autoRouteVar}->reveal());
-            $this->{$contextVar}->setAutoRoute($this->{$autoRouteVar}->reveal())->shouldBeCalled();
+        if ($route['expectedUri'] !== $route['generatedUri']) {
+            $this->uriGenerator->resolveConflict($context)->willReturn($route['expectedUri']);
         }
     }
 
     /**
-     * It should handle existing URIs.
+     * Configure the adapter.
      *
-     * @dataProvider provideBuildCollectionExisting
+     * The adapter stub:
+     *  - generates the tag,
+     *  - creates a new autoroute,
+     *  - if the route exists in the database, finds it,
+     *  - tells if the existing route matches the same content,
+     *  - if the route specify a locale, translates the content.
      */
-    public function testBuildCollectionExisting($sameContent)
+    private function configureAdapter($context, $route)
     {
-        $uri = '/uri/to';
-        $resolvedUri = '/resolved/uri';
-
-        $this->collectionBuilder->build($this->collection->reveal());
-        $this->collection->getUriContexts()->willReturn([
-            $this->context1->reveal(),
-        ]);
-        $this->collection->getSubject()->willReturn($this->subject);
-        $this->uriGenerator->generateUri($this->context1->reveal())->willReturn($uri);
-        $this->context1->setUri($uri)->shouldBeCalled();
-        $this->context1->getLocale()->willReturn(null);
-        $this->context1->getSubject()->willReturn($this->subject);
-        $this->adapter->findRouteForUri($uri, $this->context1)->willReturn(
-            $this->autoRoute1->reveal()
-        );
+        $tag = 'tag';
+        $subject = $context->reveal()->getSubject();
+
+        $this->adapter->generateAutoRouteTag($context)->willReturn($tag);
+
+        $this->adapter->createAutoRoute($context, $subject, $tag)
+            ->willReturn($this->prophesize(AutoRouteInterface::class));
 
-        // handle existing route
-        $this->adapter->compareAutoRouteContent(
-            $this->autoRoute1->reveal(),
-            $this->subject
-        )->willReturn($sameContent);
+        if ($route['existsInDatabase']) {
+            $existingAutoRoute = $this->prophesize(AutoRouteInterface::class);
 
-        $this->context1->getSubject()->willReturn($this->subject);
+            $this->adapter->findRouteForUri(
+                $route['generatedUri'],
+                $context
+            )->willReturn($existingAutoRoute);
 
-        if ($sameContent) {
-            $this->autoRoute1->setType(AutoRouteInterface::TYPE_PRIMARY)
-                ->shouldBeCalled();
+            $this->adapter->compareAutoRouteContent(
+                $existingAutoRoute->reveal(),
+                $subject
+            )->willReturn($route['withSameContent'] === true);
         } else {
```

Comment:
I meant just in terms of [calisthenics](http://williamdurand.fr/2013/06/03/object-calisthenics/#dont-use-the-else-keyword), specifically to avoid using `else`.